src/hotspot/cpu/x86/c1_LIRAssembler_x86.cpp

```diff
@@ -3514,12 +3504,22 @@ void LIR_Assembler::emit_load_klass(LIR_OpLoadKlass* op) {
   }
 
 #ifdef _LP64
-  if (UseCompressedClassPointers) {
+  if (UseCompactObjectHeaders) {
+    Register tmp = rscratch1;
+    assert_different_registers(tmp, obj);
+    assert_different_registers(tmp, result);
+
+    __ movq(result, Address(obj, oopDesc::mark_offset_in_bytes()));
+    __ shrq(result, markWord::klass_shift);
+    __ decode_klass_not_null(result, tmp);
+  } else if (UseCompressedClassPointers) {
     __ movl(result, Address(obj, oopDesc::klass_offset_in_bytes()));
     __ decode_klass_not_null(result, rscratch1);
   } else
 #endif
+  {
     __ movptr(result, Address(obj, oopDesc::klass_offset_in_bytes()));
+  }
```

Comment:
Could this if/else block be replaced with a simple call of load_klass(...)?

Reply:
Yes, will do.